controllers/imageclusterinstall_controller.go

```diff
@@ -216,15 +225,131 @@ func (r *ImageClusterInstallReconciler) Reconcile(ctx context.Context, req ctrl.
 			return ctrl.Result{}, err
 		}
 
-		if err := r.setClusterInstalledConditions(ctx, ici); err != nil {
-			log.WithError(err).Error("failed to set installed conditions")
+		// Don't check timeout or install status if cluster is already installed
+		if clusterDeployment.Spec.Installed {
+			return ctrl.Result{}, nil
+		}
+
+		timedout, err := r.checkClusterTimeout(ctx, log, ici, r.DefaultInstallTimeout)
+		if err != nil {
+			log.WithError(err).Error("failed to check for install timeout")
 			return ctrl.Result{}, err
 		}
+		if timedout {
+			log.Info("cluster install timed out")
+			return ctrl.Result{}, nil
+		}
+
+		res, err := r.checkClusterStatus(ctx, log, ici, clusterDeployment)
+		if err != nil {
+			log.WithError(err).Error("failed to check cluster status")
+			return ctrl.Result{}, err
+		}
+		return res, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ImageClusterInstallReconciler) checkClusterTimeout(ctx context.Context, log logrus.FieldLogger, ici *v1alpha1.ImageClusterInstall, defaultTimeout time.Duration) (bool, error) {
+	timeout := defaultTimeout
+
+	if timeoutOverride, present := ici.Annotations[installTimeoutAnnotation]; present {
+		var err error
+		timeout, err = time.ParseDuration(timeoutOverride)
+		if err != nil {
+			return false, fmt.Errorf("failed to parse install timeout annotation value %s: %w", timeoutOverride, err)
+		}
+	}
+
+	if ici.Status.BootTime.Add(timeout).Before(time.Now()) {
+		log.Error("timed out waiting for cluster to finish installation")
+		err := r.setClusterTimeoutConditions(ctx, ici, timeout.String())
+		if err != nil {
+			log.WithError(err).Error("failed to set cluster timeout conditions")
+		}
+		return true, err
+	}
+
+	return false, nil
+}
+
+func (r *ImageClusterInstallReconciler) checkClusterStatus(ctx context.Context, log logrus.FieldLogger, ici *v1alpha1.ImageClusterInstall, clusterDeployment *hivev1.ClusterDeployment) (ctrl.Result, error) {
+	spokeClient, err := r.spokeClient(ctx, ici)
+	if err != nil {
+		log.WithError(err).Error("failed to create spoke client")
+		return ctrl.Result{}, err
+	}
+
+	if status := r.GetSpokeClusterInstallStatus(ctx, log, spokeClient); !status.Installed {
+		log.Infof("cluster install in progress: %s", status)
+		if err := r.setClusterInstallingConditions(ctx, ici, status.String()); err != nil {
+			log.WithError(err).Error("failed to set installing conditions")
+		}
+		return ctrl.Result{RequeueAfter: time.Minute}, nil
+	}
+	log.Info("cluster is installed")
```

Comment:
should we have some cluster id in logs? I mean anything that will allow to track what happened with specific cluster

Reply:
[This](https://github.com/openshift/image-based-install-operator/blob/52a99f434b241ef89d996bdff478e81fcd141d34/controllers/imageclusterinstall_controller.go#L113) is the first thing I do in the reconcile loop.

So the ImageClusterInstall name and namespace will always be printed as fields in every log message.
Looks like this:

`time="2024-04-17T18:45:23Z" level=info msg="cluster is installed" func="github.com/openshift/image-based-install-operator/controllers.(*ImageClusterInstallReconciler).checkClusterStatus" file="/workspace/controllers/imageclusterinstall_controller.go:299" name=ibi-test namespace=ibi-test`